File: src/csharp/Foundgine.Core/Semantic/Expressions/SemanticExpression.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic.Expressions;

/// <summary>
/// Common semantic expression calculus. Expressions describe meaning only;
/// they contain no SQL, GraphQL, EF, provider, or storage instructions.
/// </summary>
public abstract record SemanticExpression
{
    public abstract SemanticType ResultType { get; }
}

public sealed record SemanticLiteralExpression(SemanticValue Value, SemanticType Type) : SemanticExpression
{
    public override SemanticType ResultType => Type;
}

public sealed record SemanticFieldReferenceExpression(FieldId Field, SemanticType Type) : SemanticExpression
{
    public override SemanticType ResultType => Type;
}

public sealed record SemanticRelationshipReferenceExpression(RelationshipId Relationship, SemanticType Type)
    : SemanticExpression
{
    public override SemanticType ResultType => Type;
}

/// <summary>A typed relationship path from a root expression to a target value.</summary>
public sealed record SemanticPathExpression(
    SemanticExpression Source,
    IReadOnlyList<RelationshipId> Relationships,
    SemanticType Type) : SemanticExpression
{
    public override SemanticType ResultType => Type;
}

public sealed record SemanticUnaryExpression(string Operator, SemanticExpression Operand, SemanticType Type)
    : SemanticExpression
{
    public override SemanticType ResultType => Type;
}

public sealed record SemanticBinaryExpression(
    string Operator,
    SemanticExpression Left,
    SemanticExpression Right,
    SemanticType Type) : SemanticExpression
{
    public override SemanticType ResultType => Type;
}

public sealed record SemanticLogicalExpression(
    SemanticLogicalOperator Operator,
    IReadOnlyList<SemanticExpression> Operands) : SemanticExpression
{
    public override SemanticType ResultType => new SemanticType.Scalar(SemanticScalarKind.Boolean);
}

public sealed record SemanticAggregateExpression(
    SemanticAggregateExpressionKind Aggregate,
    SemanticExpression Source,
    SemanticExpression? Argument = null) : SemanticExpression
{
    public override SemanticType ResultType => Aggregate switch
    {
        SemanticAggregateExpressionKind.Count => new SemanticType.Scalar(SemanticScalarKind.Int64),
        SemanticAggregateExpressionKind.Min or SemanticAggregateExpressionKind.Max
            or SemanticAggregateExpressionKind.Sum or SemanticAggregateExpressionKind.Average
            => Argument?.ResultType ?? Source.ResultType,
        _ => throw new ArgumentOutOfRangeException()
    };
}

public sealed record SemanticFunctionExpression(
    string Name,
    IReadOnlyList<SemanticExpression> Arguments,
    SemanticType Type) : SemanticExpression
{
    public override SemanticType ResultType => Type;
}

public enum SemanticLogicalOperator : byte
{
    And,
    Or
}

public enum SemanticAggregateExpressionKind : byte
{
    Count,
    Min,
    Max,
    Sum,
    Average
}

public static class SemanticExpressionTypes
{
    public static SemanticType Boolean { get; } = new SemanticType.Scalar(SemanticScalarKind.Boolean);
    public static SemanticType Int64 { get; } = new SemanticType.Scalar(SemanticScalarKind.Int64);
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/expressions/SemanticExpression.java
package com.foundgine.core.semantic.expressions;

import com.foundgine.core.abstractions.FieldId;
import com.foundgine.core.abstractions.RelationshipId;
import com.foundgine.core.semantic.*;
import java.util.*;

/** Common provider-independent semantic expression calculus. */
public interface SemanticExpression {
	SemanticType resultType();

	record Literal(SemanticValue value, SemanticType type) implements SemanticExpression {
		public SemanticType resultType() {
			return type;
		}
	}

	record FieldReference(FieldId field, SemanticType type) implements SemanticExpression {
		public SemanticType resultType() {
			return type;
		}
	}

	record RelationshipReference(RelationshipId relationship, SemanticType type) implements SemanticExpression {
		public SemanticType resultType() {
			return type;
		}
	}

	record Path(SemanticExpression source, List<RelationshipId> relationships, SemanticType type)
			implements SemanticExpression {
		public Path {
			relationships = List.copyOf(relationships);
		}

		public SemanticType resultType() {
			return type;
		}
	}

	record Unary(String operator, SemanticExpression operand, SemanticType type) implements SemanticExpression {
		public SemanticType resultType() {
			return type;
		}
	}

	record Binary(String operator, SemanticExpression left, SemanticExpression right, SemanticType type)
			implements SemanticExpression {
		public SemanticType resultType() {
			return type;
		}
	}

	record Logical(LogicalOperator operator, List<SemanticExpression> operands) implements SemanticExpression {
		public Logical {
			operands = List.copyOf(operands);
		}

		public SemanticType resultType() {
			return SemanticExpressionTypes.BOOLEAN;
		}
	}

	record Aggregate(AggregateExpressionKind aggregate, SemanticExpression source, SemanticExpression argument)
			implements SemanticExpression {
		public SemanticType resultType() {
			return switch (aggregate) {
			case COUNT -> SemanticExpressionTypes.INT64;
			case MIN, MAX, SUM, AVERAGE -> argument != null ? argument.resultType() : source.resultType();
			};
		}
	}

	record Function(String name, List<SemanticExpression> arguments, SemanticType type) implements SemanticExpression {
		public Function {
			arguments = List.copyOf(arguments);
		}

		public SemanticType resultType() {
			return type;
		}
	}

	enum LogicalOperator {
		AND, OR
	}

	enum AggregateExpressionKind {
		COUNT, MIN, MAX, SUM, AVERAGE
	}

	final class SemanticExpressionTypes {
		private SemanticExpressionTypes() {
		}

		public static final SemanticType BOOLEAN = new SemanticType.Scalar(SemanticScalarKind.BOOLEAN);
		public static final SemanticType INT64 = new SemanticType.Scalar(SemanticScalarKind.INT64);
	}
}
